### PR TITLE
feat: add logging through Weights & Biases

### PR DESCRIPTION
**Pull request by borisdayma**
Hi,

I've been using hugging-face for a while and I've been loving it. I think it could become a central reference for NLP problems and the new `Trainer` class made it so much easier to add new functionality.

I would like to contribute in solving a few limitations I observed:
* some models don't have clear details on how they have been trained (hyper-parameters) and for which task, making it hard to reproduce their results (and even ensure those results were actually reached) -> ideally we would track & log their training process as well as config parameters used
* some new models keep on being requested/proposed without seeing their actual efficiency -> ideally we would be able to compare them to baselines
* I've seen issues where "better" models are offered to replace previous ones but there is no way to prove they are actually better -> maybe we can add a metrics table (linked to actual runs) to the model card
* it is hard to keep track of models and datasets
* overall I'd like to bring more traceability and transparency between users

This is why I propose to use [Weights & Biases](https://docs.wandb.com/) which is free for open source and will help solve those issues. Also it is already part of Pytorch-Lightning so it will be easy to integrate it with those scripts, if extra setup is required.

This PR brings :

* logging through Weights & Biases when `wandb` is installed
* logging of evaluation metrics at the end of training

See example run on distilbert-base-cased  -> [W&B run](https://app.wandb.ai/borisd13/transformers-examples/runs/3iuvnwam?workspace=user-borisd13)

image

If you navigate around, you will see that config parameters, metrics, gradients, computer resources, actual code (including git repo & command line used) and stdout have all been uploaded for full traceability and reproducibility. I can also easily upload trained model but maybe we could add an arg to make it optional in case users have limited connection (as you prefer)?

I can also easily compare bert-base-uncased and distilbert-base-uncased from my [project page](https://app.wandb.ai/borisd13/transformers-examples).
*Note: I didn't try any fine-tuning and just used same parameters on both in this example.*

image

This makes it easy to quickly see if a new model is actually interesting.

Finally, this integration let us create [cool reports](https://app.wandb.ai/stacey/keras_finetune/reports/Curriculum-Learning-in-Nature--Vmlldzo1MjcxNw) where graphs directly pool data and are interactive.

If you like it, then I'd love to add other features in later PR's:

* log trained model -> I'll just need to have access to path of relevant files so that I don't upload entire output directory
* add [W&B init args](https://docs.wandb.com/library/init) as parameters based on what can be useful for hugging-face
* log all command line args -> I need to access the parser so that I can also log parameters such as `model_path_or_name` and `max_seq_length` (for run_glue script) which will be convenient to create custom summary tables and compare models (or finetune)
* maybe log called script (unless `finetuning_taks` is enough?)
* in addition to metrics, track model size & inference speed which can help users choose the model they want
* use artifacts (W&B version control system) to track models, tokenizers & datasets -> central repository that will let us see easily that runs are based on same dataset or same pre-trained model + see all trained tasks associated to one pre-trained model
* log prediction samples (based on the task)
* create a central repo on W&B (entity: hugging-face) to have access to other's people runs
* integrate model card issue template so that we always add a link to a run as well as performance metrics
* add [sweeps](https://docs.wandb.com/sweeps) -> make it easy to optimize hyper-parameters
* we could add a hook to run automatically on all tasks any new model added (and link the run to the model cards)

Please let me know your thoughts and feel free to contact me or Weights & Biases directly to discuss this integration.

### CODE REVIEW

**Comment by borisdayma** [comment]
Hi,
I'm just checking if there's anything you would like me to change on this PR.
Ideally I would be able to pass a few extra optional kwargs (saving model, project name…) but the integration can go as deep as you want based on the project philosophy.

The idea is mainly to give more traceability to models and a convenient way to compare and optimize models.

**Comment by julien-c** [comment]
Will take a look very soon @borisdayma! (can i ping you on the W&B Slack or is there a better channel?)

**Comment by borisdayma** [comment]
Perfect, you can ping me on the W&B slack!

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -220,6 +225,12 @@ def get_optimizers(
         )
         return optimizer, scheduler
 
+    def _setup_wandb(self):
+        # Start a wandb run and log config parameters
+        wandb.init(name=self.args.logging_dir, config=vars(self.args))
+        # keep track of model topology and gradients
+        # wandb.watch(self.model)
```
Comment (julien-c) [inline src/transformers/trainer.py:232]: This line caused a crash for me later on so I commented it out @borisdayma 

In any case, one can override this method to customize the wandb initialization if needed.

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -368,14 +376,8 @@ def train(self, model_path: Optional[str] = None):
...
-                            if _has_wandb:
-                                wandb.log(
-                                    {
-                                        **logs,
-                                        "epoch": global_step
-                                        / (len(train_dataloader) // self.args.gradient_accumulation_steps),
-                                    }
-                                )
+                            if is_wandb_available():
+                                wandb.log(logs, step=global_step)
```
Comment (julien-c) [inline src/transformers/trainer.py:380]: did a simpler version of this

File: src/transformers/trainer.py [hunk 3]
```diff
@@ -515,11 +514,6 @@ def evaluate(
         eval_dataloader = self.get_eval_dataloader(eval_dataset)
 
         output = self._prediction_loop(eval_dataloader, description="Evaluation")
-        if _has_wandb and not no_log and wandb.run:
-            # Log final metrics
-            wandb.log(
-                {**{"eval_{}".format(k): v for k, v in output.metrics.items()}, "epoch": self.args.num_train_epochs}
-            )
```
Comment (julien-c) [inline src/transformers/trainer.py]: If one needs this, probably simpler to just do it from the calling script when you have the eval results

File: src/transformers/trainer.py (hunk 1 above)
Comment (borisdayma) [reply src/transformers/trainer.py:232]: What it does is logging gradients and model topology as in [this run](https://app.wandb.ai/borisd13/transformers-examples/runs/3iuvnwam?workspace=user-borisd13). It's not essential but can be interesting.
Do you have an example of where it failed?
Note: this needs to be called after `wandb.init`.

File: src/transformers/trainer.py (hunk 2 above)
Comment (borisdayma) [reply src/transformers/trainer.py:380]: The only purpose here was to log `epoch` (as a float). When you press the x-axis, you can use "epoch" instead of just the "global_step".
I find it especially useful when I want to compare 2 runs with different batch size for example (let's say if a model is much bigger than another and I cannot fit as many samples in my GPU).

File: src/transformers/trainer.py (hunk 3 above)
Comment (borisdayma) [reply src/transformers/trainer.py]: Makes sense + it simplifies the trainer

**Comment by borisdayma** [comment]
That's great @julien-c !
The next big piece that could be useful is to log the trained model weights on W&B.
I am not sure it's a good idea to do it automatically. Would there be an easy way to add an arg for it `save_model`?

File: src/transformers/trainer.py (hunk 3 above)
Comment (julien-c) [reply src/transformers/trainer.py]: (personally I always use the `--evaluate_during_training` flag so I don't really need it here anyways)

File: src/transformers/trainer.py (hunk 2 above)
Comment (julien-c) [reply src/transformers/trainer.py:380]: Makes sense. Let's look for a way to add it back in a subsequent PR (would be useful for TensorBoard too)

File: src/transformers/trainer.py (hunk 1 above)
Comment (julien-c) [reply src/transformers/trainer.py:232]: Tried to reproduce today (and open an issue on https://github.com/wandb/client) but for some reason, couldn't – so I'll uncomment this.

**Comment by julien-c** [comment]
I'll add some doc about wandb to our examples' README.md

File: src/transformers/trainer.py (hunk 2 above)
Comment (julien-c) [reply src/transformers/trainer.py:380]: uncommented in fd2174664c8879c747ada3e6e0a2486858808421

**Comment by borisdayma** [comment]
Awesome, let me know if I can be of any further help!

File: src/transformers/trainer.py (hunk 3 above)
Comment (borisdayma) [reply src/transformers/trainer.py]: I just remembered the use. It was too get final evaluation results at last training step (which may not be a logging step otherwise).

In those cases I works just log evaluation metrics (not the training ones).

On the runs I sent, we can see the metrics are logged at the end (but no training loss).

The idea is to have final performance on the model. What do you think?
